noir-projects/noir-contracts/contracts/fpc_contract/src/main.nr

```diff
@@ -1,76 +1,120 @@
+    /// to pay is defined by `max_fee` and is denominated in AA.
+    ///
+    /// ## Overview
+    /// The refund flow works as follows:
+    /// 1. We pull the `max_fee` from the user's balance of the AA to this contract (requires setting an authwit),
+    /// 2. then the private and public functions of a tx get executed,
+    /// 3. at this point we know the tx fee so we can compute how much of AA the user needs to pay to
+    /// the `private_fee_recipient` and how much of it will be refunded back. Note that this is computed based on
+    /// an exchange rate between AA and fee juice.
+    /// 4. the protocol deducts the actual fee denominated in fee juice from the FPC's balance.
+    ///
+    /// ***Note:***
+    /// AA funds sent by the users to this contract stay in this contract and are not transferred
+    /// to the `private_fee_recipient`. In the private flow we needed a separate `private_fee_recipient` as this
+    /// contract does not have keys associated with it. In production we would want to have a method allowing for
+    /// pulling of these funds from this contract.
     #[private]
-    fn fee_entrypoint_public(amount: Field, asset: AztecAddress, nonce: Field) {
-        FPC::at(context.this_address())
-            .prepare_fee(context.msg_sender(), amount, asset, nonce)
+    fn fee_entrypoint_public(max_fee: Field, nonce: Field) {
+        // TODO(PR #8022): Once PublicImmutable performs only 1 merkle proof here, we'll save ~4k gates
+        let config = storage.config.read();
+
+        // We pull the max fee from the user's balance of the accepted asset to this contract.
+        // docs:start:public_call
+        Token::at(config.accepted_asset)
+            .transfer_in_public(context.msg_sender(), context.this_address(), max_fee, nonce)
             .enqueue(&mut context);
+        // docs:end:public_call
+
         context.set_as_fee_payer();
         // TODO(#6277) for improving interface:
-        // FPC::at(context.this_address()).pay_refund(context.msg_sender(), amount, asset).set_public_teardown_function(&mut context);
+        // FPC::at(context.this_address()).pay_refund(...).set_public_teardown_function(&mut context);
         context.set_public_teardown_function(
             context.this_address(),
             comptime { FunctionSelector::from_signature("pay_refund((Field),Field,(Field))") },
-            [context.msg_sender().to_field(), amount, asset.to_field()],
+            [context.msg_sender().to_field(), max_fee, config.accepted_asset.to_field()],
         );
     }
 
+    /// Pays the refund to the `refund_recipient`. The refund is the difference between the `max_fee` and
+    /// the actual fee. `accepted_asset` is the asset in which the refund is paid. It's passed as an argument
+    /// to avoid the need for another read from public storage.
     #[public]
     #[internal]
-    fn prepare_fee(from: AztecAddress, amount: Field, asset: AztecAddress, nonce: Field) {
```

Comment:
Nuked this function as it was superfluous --> we could just enqueue the call to the token directly from the private function (which saves 1 kernel iteration).